src/librustc/mir/interpret/allocation.rs

```diff
@@ -41,6 +45,481 @@ pub struct Allocation<Tag=(),Extra=()> {
+        match self.bytes[offset..].iter().position(|&c| c == 0) {
+            Some(size) => {
+                let p1 = Size::from_bytes((size + 1) as u64);
+                self.check_relocations(cx, ptr, p1)?;
+                self.check_defined(ptr, p1)?;
+                Ok(&self.bytes[offset..offset + size])
+            }
+            None => err!(UnterminatedCString(ptr.erase_tag())),
+        }
+    }
+
+    pub fn check_bytes(
+        &self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        size: Size,
+        allow_ptr_and_undef: bool,
+    ) -> EvalResult<'tcx> {
+        let align = Align::from_bytes(1, 1).unwrap();
+        // Check bounds, align and relocations on the edges
+        self.get_bytes_with_undef_and_ptr(cx, ptr, size, align)?;
+        // Check undef and ptr
+        if !allow_ptr_and_undef {
+            self.check_defined(ptr, size)?;
+            self.check_relocations(cx, ptr, size)?;
+        }
+        Ok(())
+    }
+
+    pub fn read_bytes(
+        &self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        size: Size,
+    ) -> EvalResult<'tcx, &[u8]> {
+        let align = Align::from_bytes(1, 1).unwrap();
+        self.get_bytes(cx, ptr, size, align)
+    }
+
+    pub fn write_bytes(
+        &mut self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        src: &[u8],
+    ) -> EvalResult<'tcx> {
+        let align = Align::from_bytes(1, 1).unwrap();
+        let bytes = self.get_bytes_mut(
+            cx, ptr, Size::from_bytes(src.len() as u64), align,
+        )?;
+        bytes.clone_from_slice(src);
+        Ok(())
+    }
+
+    pub fn write_repeat(
+        &mut self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        val: u8,
+        count: Size
+    ) -> EvalResult<'tcx> {
+        let align = Align::from_bytes(1, 1).unwrap();
+        let bytes = self.get_bytes_mut(cx, ptr, count, align)?;
+        for b in bytes {
+            *b = val;
+        }
+        Ok(())
+    }
+
+    /// Read a *non-ZST* scalar
+    pub fn read_scalar(
+        &self,
+        cx: &impl HasDataLayout,
+        ptr: Pointer<Tag>,
+        ptr_align: Align,
+        size: Size
+    ) -> EvalResult<'tcx, ScalarMaybeUndef<Tag>> {
+        // get_bytes_unchecked tests alignment and relocation edges
+        let bytes = self.get_bytes_with_undef_and_ptr(
+            cx, ptr, size, ptr_align.min(self.int_align(cx, size))
+        )?;
+        // Undef check happens *after* we established that the alignment is correct.
+        // We must not return Ok() for unaligned pointers!
+        if self.check_defined(ptr, size).is_err() {
+            // this inflates undefined bytes to the entire scalar, even if only a few
+            // bytes are undefined
+            return Ok(ScalarMaybeUndef::Undef);
+        }
+        // Now we do the actual reading
+        let bits = read_target_uint(cx.data_layout().endian, bytes).unwrap();
```

Comment:
if `bytes` has zero elements, we'd transitively hit the assertion in https://docs.rs/byteorder/1.2.7/src/byteorder/lib.rs.html#1905

Reply:
Okay. Please add a comment here saying that this is why ZST don't work, and maybe also that this is no problem because for ZST we anyway (already in `Memory`) cannot convert a `Scalar` to a `Pointer`.